src/core/toPath.js

```diff
@@ -102,139 +110,84 @@ const allowingArrays = fn => arg => {
-
-        // Shorthand: if array index is the whole slice add it to path
-        if (arrayIndexValue === ':') {
-          path.push([undefined, undefined])
-        } else {
-
-          // Look for a slice quote
-          const sliceDelimIndex = arrayIndexValue.indexOf(':')
-
-          // If no slice quote found
-          if (sliceDelimIndex === -1) {
-            // Parse array index as a number
-            const nArrayIndexValue = Number(arrayIndexValue)
-
-            // Add array index to path, either as a valid index (positive int), or as a string
-            path.push(isIndex(nArrayIndexValue) ? nArrayIndexValue : arrayIndexValue)
-
-          } else { // If a slice quote is found
-
-            // Fetch slice start and end, and parse them as slice indexes (empty or valid int)
-            const sliceStart = arrayIndexValue.substring(0, sliceDelimIndex), sliceEnd = arrayIndexValue.substring(sliceDelimIndex + 1)
-            const nSliceStart = toSliceIndex(sliceStart), nSliceEnd = toSliceIndex(sliceEnd)
-
-            // Add array index to path, as a slice if both slice indexes are valid (undefined or int), or as a string
-            path.push(isSliceIndex(nSliceStart) && isSliceIndex(nSliceEnd) ? [nSliceStart, nSliceEnd] : arrayIndexValue)
-          }
-        }
+  const path = stringToPath2(str)
+  return str[0] === '.' ? ['', ...path] : path
+}
 
-        // Stop if end of string has been reached
-        if (index === str.length) break
-      }
-    }
+/**
+ * @typedef {function(string): string[]} Matcher a function that can replace String.prototype.match
+ * @param {string} str string to match against
+ * @param {[(Matcher | RegExp), function(string[]): *]} matchers
+ *   pairs of a regexp to match str against, and a function to transform the resulting match object into the final result
+ * @param {*} defaultResult
+ *   value to return if no matcher matches
+ * @returns {*} output value of the first cond that matches or defaultResult if no cond matches
+ */
+const match = (str, matchers, defaultResult) => {
+  for (const [matcher, mapper] of matchers) {
+    const match = matcher instanceof RegExp ? str.match(matcher) : matcher(str)
+    if (match) return mapper(match.slice(1))
+  }
+  return defaultResult
+}
 
+match.andCheck = (matcher, predicate) => {
+  return str => {
+    const match = str.match(sliceNotation)
+    if (!match) return match
+    return predicate(match.slice(1)) ? match : null
   }
+}
 
-  return path
+const quotedBracketNotation = /^\[(['"])(.*?[^\\])\1\]?\.?(.*)$/
+const incompleteQuotedBracketNotation = /^\[["'](.*)$/
+const sliceNotation = /^\[([^:\]]*):([^:\]]*)\]\.?(.*)$/
+const bareBracketNotation = /^\[([^\]]*)\]\.?(.*)$/
+const incompleteBareBracketNotation = /^\[(.*)$/
+const pathSegmentEndedByDot = /^([^.[]*?)\.(.*)$/
+const pathSegmentEndedByBracket = /^([^.[]*?)(\[.*)$/
+
+const stringToPath2 = str => {
+  return match(str, [
+    [
+      str => str.length === 0 ? [] : null,
```

Comment:
As a reader, this just looks weird until I understand how `match` works. But I don't know how to make it better.

Reply:
Maybe an object ? 
```js 
{
  matcher: ...,
  mapper: ...,
}
```

Reply:
I like the idea, though I'm not sure that would be the ideal key names. What about

```
{
  ifMatch: ..., // or maybe simply 'match'
  then: ...,
}
```

I feel this expresses the intent.

Reply:
What do you think if we create a new function `makeMatcher(matcher, mapper)` which would create a new function returning either null or the result of the mapper.
Then match would just send the first non null match... or the default value.

Reply:
Sounds nicer. More composable. Parser-combinator-like.

Reply:
Working on it.